packages/dds/tree/src/feature-libraries/modular-schema/modularChangeFamily.ts

```diff
@@ -549,7 +576,28 @@ export class ModularChangeFamily
 		genId: IdAllocator,
 		crossFieldTable: CrossFieldTable<RebaseData>,
 		revisionMetadata: RevisionMetadataSource,
+		constraintState: ConstraintState,
 	): NodeChangeset {
+		// We only care if a violated constraint is fixed or if a non-violated
+		// constraint becomes violated
+		if (
+			change.valueConstraint !== undefined &&
+			over.change.valueChange !== undefined &&
+			over.change.valueChange.value !== change.valueConstraint.value &&
+			change.valueConstraint.violated !== true
+		) {
+			change.valueConstraint.violated = true;
+			constraintState.violationCount++;
+		} else if (
+			change.valueConstraint !== undefined &&
+			over.change.valueChange !== undefined &&
+			over.change.valueChange.value === change.valueConstraint.value &&
+			change.valueConstraint.violated === true
+		) {
+			change.valueConstraint.violated = false;
+			constraintState.violationCount--;
+		}
```

Comment:
```suggestion
		if (
			change.valueConstraint !== undefined &&
			over.change.valueChange !== undefined
		) {
			const violatedByOver = over.change.valueChange.value !== change.valueConstraint.value;
			if (!change.valueConstraint.violated && violatedByOver ) {
				change.valueConstraint.violated = true;
				constraintState.violationCount++;
			} else if (change.valueConstraint.violated && !violatedByOver ) {
				change.valueConstraint.violated = false;
				constraintState.violationCount--;
			}
		}
```

Reply:
Nice, this block was super ugly. Much nicer with these changes.